### PR TITLE
fix pilot goroutine leak

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
fixes: #10822

replace: #10862

### CODE REVIEW

File: pilot/pkg/proxy/envoy/v2/ads.go [hunk 1]
```diff
@@ -358,7 +358,13 @@ func receiveThread(con *XdsConnection, reqChannel chan *xdsapi.DiscoveryRequest,
 			totalXDSInternalErrors.Add(1)
 			return
 		}
-		reqChannel <- req
+		select {
+		case <-con.stream.Context().Done():
```
Comment (qfel) [inline pilot/pkg/proxy/envoy/v2/ads.go]: Hmm, there's also con.doneChannel, which depending on what happens is closed slightly after or slightly before the context is done. I wonder if we should just get rid of it and always use Context().Done()?

BTW there's also https://github.com/istio/istio/pull/10862

Comment (hzxuzhonghu) [reply pilot/pkg/proxy/envoy/v2/ads.go]: of course, we can.

**Comment by pxzero** [comment]
i have try this pr , but still have the same go routine leak problem . i think the trouble is this code 
```go
                select {
		case <-con.stream.Context().Done():
		        adsLog.Errorf("ADS: %q %s terminated with stream closed", con.PeerAddr, con.ConID)
			return
		default:
			reqChannel <- req
		}
```
it should change like this,so that receiveThread will not block at reqChannel <- req and handle grpc stream close correctly
```go
case reqChannel <- req:
```
**BTW , Could we add a test case for this scenes? i think this bug is dangerous.**

**Comment by hzxuzhonghu** [comment]
This is not right, 　Ｉ　fixed it in #10998

**Comment by hzxuzhonghu** [comment]
Of course, we can add a test. Will do when I am free.

**Comment by pxzero** [comment]
> This is not right, 　Ｉ　fixed it in #10998

greet

**Pull request merged**